---
summary: 
locale: en-us
guid: cba49836-50fc-4505-a2ea-c5e729c20580
app_type: mobile apps, reactive web apps
platform-version: odc
figma:
---

<h1>Consuming an API from the Private Gateway results in a "404 not found" Response</h1>
<p>
<br/>
</p>
<p>Error Code: OS-BERT-00000</p>
<p>
<br/>
</p>
<p>
<strong>Symptoms</strong>: 404 Error - Not Found, Unable to consume REST API</p>
<h2>Precautions</h2>
<p>This error covers a relatively specific but common scenario where consuming an API via the Private Gateway results in a "404 not found" response, but it is not meant to be a comprehensive list of the potential causes for this error.</p>
<p>
<br/>
</p>
<h2>Troubleshooting</h2>
<p>You may notice that, when you <a href="https://success.outsystems.com/documentation/outsystems_developer_cloud/managing_outsystems_platform_and_apps/configure_a_private_gateway_to_your_network/#use-endpoints-in-your-apps" target="_blank" rel="noopener noreferrer">consume an API through the Private Gateway</a>, the method will return a "404 Not Found" reponse, similar to the example below. If so, you can also check the Traces to ensure this response is coming from the API you are consuming:</p>
<p>
![Error executing Get with 404 Not Found message in the OutSystems platform.](images/im-image-ck-2025225rk0U7TqQdJVaxghqbiMPFUbiN8O8bdTPbaHvyO2oZz.png "Error executing Get - 404 Not Found")
</p>
<p>
![Trace details showing 404 error for GET request in OutSystems platform.](images/im-image-ck-2025225lcqGxF68ipI0ZYtrvuf86NP5URxZhcypnXeWDfOFJe.png "API Trace Details")
</p>
<p>There may be multiple causes for this error, including an issue in the API you are consuming or connection issues from your Cloud Connector. That said, if you are getting the error above, please open the app/library where you are consuming this API in ODC Studio and confirm if you have an "OnBeforeRequest" API Callback on it like in the screenshot below:</p>
<p>
![OnBeforeRequest API Callback configuration in OutSystems Studio.](images/im-image-ck-2025225eVtxKuZkbbGIn3XQMLUJlioIpJ1WrFbQs8kNyZ8FLM.png "OnBeforeRequest API Callback")
</p>
<p>If you do not have this Callback, proceed to Incident Resolution Measures; if you do, you may still proceed to Incident Resolution Measures to confirm if it has the logic we'll suggest in the section below.</p>
<h2>Incident Resolution Measures</h2>
<p>As implied in the Troubleshooting section, it's possible that the error is caused by a lack of "OnBeforeRequest" API Callback; the reason for this is that, for many APIs, they require their own URL to be in the Header of the request when you try to consume them. The issue with this is that, in ODC Studio, the "URL" that the request is going to is "secure-gateway:[port]", which may cause the API to reject the request.</p>
<p>By adding an OnBeforeRequest, you can edit the header to specifically include the API's URL in the "Host" value as required; the value should be the api's hostname (which you probably used in the command to execute <a href="https://github.com/OutSystems/cloud-connector" target="_blank" rel="noopener noreferrer">Cloud Connector</a>). As such, perform these actions:</p>
<ul>
    <li>Add a new Action OnBeforeRequest:<ul>
    <li>
        ![Properties of Securegateway REST API in OutSystems Studio.](images/im-image-ck-20252254ODuXS4GlySQ4kEawQYtwVZhKoWobPYNJWO9RVR9EI.png "Securegateway REST API Properties")
        </li>
    </ul>
</li>
<li>Drag the Server Action <strong>ListAppend </strong>to the OnBeforeRequest's logic before the End and set its parameters as follows:<ul>
<li>
    <p>List: CustomizedRequest.Headers</p>
</li>
<li>
    <p>Element.Name: "Host"</p>
</li>
<li>
    <p>Element.Value: "[API URL]"</p>
</li>
<li>
    <p>
    ![ListAppend server action configuration in OnBeforeRequest logic in OutSystems Studio.](images/im-image-ck-2025224GUvZJVQf06flENOOTB9MRCV85a3ajF8UZZ016BrYVB.png "ListAppend Server Action")
    </p>
</li>
</ul>
</li>
</ul>
<p>If you already have an OnBeforeRequest with other necessary logic, just ensure it includes this as well.</p>
<p>If you are still running into the same response after performing these steps, please confirm if all the information is correct, including the API's URL and the connection through Cloud Connector; if everything looks correct and the issue still persists, please <a href="https://success.outsystems.com/support/home/" target="_blank" rel="noopener noreferrer">contact our Global Support team</a>.</p>